Override construction of DeploymentExceptions for RM Startup

abutch3r opened the pull request:
Override ReactiveMessagingExtension startup process to not start the  DeploymentException that is thrown to indicate wiring exceptions.

Update Reactive packages to latest
Mutiny - 2.2.0 -> 2.5.1
Reactive Messaging - 4.8.0 -> 4.11.0
Mutiny Convertor Api - 3.0.0 -> 3.0.1

Update EmitterImpl to match latest RM Provider version due to changes around blocking Update WorkerPoolRegistry to include properties now used in MediatorManager

On error we now get errors in the messages.log such as 
```
[27/10/2023, 14:01:13:566 BST] 00000210 reactive.messaging30.internal.OLReactiveMessaging30Extension E Some components are not connected to either downstream consumers or upstream producers:
	- IncomingConnector{channel:'dummy-source-2', attribute:'mp.messaging.incoming.dummy-source-2'} has no downstream
	- IncomingConnector{channel:'dummy-source', attribute:'mp.messaging.incoming.dummy-source'} has no downstream
	- ProcessingMethod{method:'org.eclipse.microprofile.reactive.messaging.tck.connector.MyProcessorWithBadStreamName#process', incoming:'bad name', outgoing:'dummy-sink'} has no upstream

[27/10/2023, 14:01:13:584 BST] 00000210 com.ibm.ws.app.manager.AppMessageHelper                      E CWWKZ0002E: An exception occurred while starting the application test-c8d284f3-f976-432b-9fa5-2f51aca9a846. The exception message was: com.ibm.ws.container.service.state.StateChangeException: org.jboss.weld.exceptions.DeploymentException: Wiring error(s) detected in application. Review the server messages.log file and FFDC logs to identify the problem(s).
```

Fixes #26664
Some Fixes for #26617

abutch3r committed:
Override construction of DeploymentExceptions for Reactive Messaging startup

Override MediatorManager startup process to change the construction of the DeploymentException that is thrown to indicate wiring exceptions.

Update Reactive packages to latest
Mutiny - 2.2.0 -> 2.5.1
Reactive Messaging - 4.8.0 -> 4.10.1
Mutiny Convertor Api - 3.0.0 -> 3.0.1

Update EmitterImpl to match latest RM Provider version due to changes around blocking
Update WorkerPoolRegistry to include properties now used in MediatorManager

abutch3r commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.reactive.messaging_fat,com.ibm.ws.microprofile.reactive.messaging_fat_tck,com.ibm.ws.microprofile.reactive.streams.operators_fat,com.ibm.ws.microprofile.reactive.streams.operators_fat_tck,io.openliberty.microprofile.reactive.messaging30.internal_fat_tck,io.openliberty.microprofile.reactive.messaging.internal_fat,io.openliberty.microprofile.reactive.streams.operators30.internal_fat_tck

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/smallrye/reactive/messaging/providers/extension/ReactiveMessagingExtension.java:
```diff
@@ -0,0 +1,271 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ******************************************************************************/
```
Unsure about the copyright header on this. This is a smallrye class which we've modified, I don't think it should just have an IBM copyright header on it.

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/smallrye/reactive/messaging/providers/extension/ReactiveMessagingExtension.java:
```diff
@@ -0,0 +1,271 @@
+import jakarta.enterprise.inject.spi.WithAnnotations;
+import jakarta.inject.Inject;
+
+/**
+ *
+ */
+public class ReactiveMessagingExtension implements Extension {
```
This should have a better header saying where it's from and that we've modified it.

The comments below around the actual modifications are good.

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java:
```diff
@@ -87,6 +94,24 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
+        // Catch DeploymentException that is thrown if there are any validation errors with the Reactive Messaging Configuration
+        try {
+           mediatorManager.start();
+        } catch (DeploymentException de){
+            for(Throwable e: de.getSuppressed()){
```
Code not formatted

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java:
```diff
@@ -87,6 +94,24 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
+                Tr.error(tc, e.getLocalizedMessage());
+            }
+            // throw new DeploymentException to ensure application stops with the right exception and improve the message
+            throw new DeploymentException("Wiring error(s) detected in application. Review the server messages.log file and FFDC logs to identify the problem(s).", de);
```
If we're throwing a new exception with a new message, the message should be translated.

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java:
```diff
@@ -87,6 +94,24 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
+            // throw new DeploymentException to ensure application stops with the right exception and improve the message
+            throw new DeploymentException("Wiring error(s) detected in application. Review the server messages.log file and FFDC logs to identify the problem(s).", de);
+        }
+    }
```
In general, I'm unsure about logging the exceptions as their own log messages without any additional message and then logging the app start failure.

Usually we'd build up one long message with newlines which would be appended to a translated message.

Also, are we certain that a `DeploymentException` always means wiring errors?

abutch3r committed:
Make requested updates

Include NLS bundle for our message for when an error occurs
Improve, building of errors string
Fix copyright

abutch3r commented:
#libby

abutch3r commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.reactive.messaging_fat,com.ibm.ws.microprofile.reactive.messaging_fat_tck,com.ibm.ws.microprofile.reactive.streams.operators_fat,com.ibm.ws.microprofile.reactive.streams.operators_fat_tck,io.openliberty.microprofile.reactive.messaging30.internal_fat_tck,io.openliberty.microprofile.reactive.messaging.internal_fat,io.openliberty.microprofile.reactive.streams.operators30.internal_fat_tck

dmuelle commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/resources/io/openliberty/microprofile/reactive/messaging30/resources/ReactiveMessagingProvider.nlsprops:
```diff
@@ -0,0 +1,32 @@
+# CWMRX   1000-1999       MicroProfile Reactive Messaging
+#-----------------------------------------------------------------------------------------------------------------------------
+
+reactive.messaging.validation.error.CWMRX1100E=CWMRX1100E: An error occurred during Reactive Messaging validation while starting the {0} application. The error was {1}
```
```suggestion
reactive.messaging.validation.error.CWMRX1100E=CWMRX1100E: An error occurred during Reactive Messaging validation when the {0} application started. The error was {1}
```

Azquelt replied:
Thanks David. The error message for `{1}` here will usually be quite long (multiple lines) so I don't think the final period should be included. Is that ok?

Otherwise I agree with the change, thank you.

dmuelle replied:
Yes- thanks Andrew, I updated my suggestion to remove it.

dmuelle commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/resources/io/openliberty/microprofile/reactive/messaging30/resources/ReactiveMessagingProvider.nlsprops:
```diff
@@ -0,0 +1,32 @@
+#-----------------------------------------------------------------------------------------------------------------------------
+
+reactive.messaging.validation.error.CWMRX1100E=CWMRX1100E: An error occurred during Reactive Messaging validation while starting the {0} application. The error was {1}
+reactive.messaging.validation.error.CWMRX1100E.explanation=During validation of the constructed Reactive Messaging graph, errors have been found that would prevent the ability of sending or receiving messages.
```
```suggestion
reactive.messaging.validation.error.CWMRX1100E.explanation=During validation of the constructed Reactive Messaging graph, errors were found that prevent the sending or receiving of messages.
```

dmuelle reviewed (changes requested):
ID review

abutch3r committed:
Make ID Requested updates

abutch3r commented:
#libby

abutch3r commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.reactive.messaging_fat,com.ibm.ws.microprofile.reactive.messaging_fat_tck,com.ibm.ws.microprofile.reactive.streams.operators_fat,com.ibm.ws.microprofile.reactive.streams.operators_fat_tck,io.openliberty.microprofile.reactive.messaging30.internal_fat_tck,io.openliberty.microprofile.reactive.messaging.internal_fat,io.openliberty.microprofile.reactive.streams.operators30.internal_fat_tck

dmuelle reviewed (dismissed):
LGTM

abutch3r commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.reactive.messaging_fat,com.ibm.ws.microprofile.reactive.messaging_fat_tck,com.ibm.ws.microprofile.reactive.streams.operators_fat,com.ibm.ws.microprofile.reactive.streams.operators_fat_tck,io.openliberty.microprofile.reactive.messaging30.internal_fat_tck,io.openliberty.microprofile.reactive.messaging.internal_fat,io.openliberty.microprofile.reactive.streams.operators30.internal_fat_tck

posmith commented:
L2 message review completed.

abutch3r committed:
Update to RM 4.11.0

benjamin-confino commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java:
```diff
@@ -87,6 +95,23 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
     }
 
+    // Set priority to be higher then default, so will run after the ReactiveMessagingExtension event
+    void postDeploymentValidation(@Observes @Priority(jakarta.interceptor.Interceptor.Priority.APPLICATION + 1000) AfterDeploymentValidation done, BeanManager beanManager) {
```
If you change this method signature so it overrides `protected void afterDeploymentValidation(@Observes AfterDeploymentValidation done, BeanManager beanManager)` from the superclass and add `MediatorManager mediatorManager = configureMediatorManager(beanManager);` to the start of the try block, I think you can remove the overlay entirely.

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/src/io/openliberty/microprofile/reactive/messaging30/internal/OLReactiveMessaging30Extension.java:
```diff
@@ -87,6 +95,24 @@ void beforeBeanDiscovery(@Observes BeforeBeanDiscovery discovery, BeanManager be
+        try {
+            MediatorManager mediatorManager = configureMediatorManager(beanManager);
+            startMediatorManager(mediatorManager);
+        } catch (DeploymentException de) {
+            StringBuilder exceptionBuilder = new StringBuilder(de.getLocalizedMessage());
+            for (Throwable t : de.getSuppressed()) {
+                exceptionBuilder.append("\n" + t.getLocalizedMessage());
+            }
+            String appName = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData().getJ2EEName().getApplication();
+            Tr.error(tc,  Tr.formatMessage(tc, "reactive.messaging.validation.error.CWMRX1100E", appName, exceptionBuilder.toString()));
```
We should have a test for the error message we're outputting here.

Azquelt replied:
Noted message above, see #26778 for new tests.

abutch3r commented:
Additional tests that cover the added error messages and contents will be added under https://github.com/OpenLiberty/open-liberty/pull/26778

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/bnd.bnd:
```diff
@@ -43,13 +43,16 @@ IBM-Default-Config: OSGI-INF/wlp/defaultInstances.xml
 Private-Package: \
 	io.smallrye.reactive.messaging.providers.connectors;-split-package:=merge-first,\
 	io.smallrye.reactive.messaging.*,\
-	io.openliberty.microprofile.reactive.messaging30.internal
+	io.openliberty.microprofile.reactive.messaging30.internal,\
+	io.openliberty.microprofile.reactive.messaging30.resources;version=1.0
```
1. There's no need to version packages that aren't exported
2. This package also has a `package-info.java` with a version defined. You can either put the version here in the bnd, or in the `package-info.java`, you don't want it in both because they can then get out of sync.

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/bnd.bnd:
```diff
@@ -59,7 +62,9 @@ Private-Package: \
   io.openliberty.io.smallrye.reactive.converter-api3;version=latest,\
-  io.openliberty.org.jboss.logging35;version=latest
+  io.openliberty.org.jboss.logging35;version=latest,\
+  jakarta.enterprise.cdi-api;version=latest
```
You shouldn't need this dependency because you already have `io.openliberty.jakarta.cdi.3.0;version=latest` above.

In general we prefer to depend on liberty's repackaged bundles of jakarta APIs.

abutch3r replied:
it does build fine without, so will include a future PR before shipping

Azquelt commented on dev/io.openliberty.io.smallrye.reactive.messaging-provider4/bnd.bnd:
```diff
@@ -59,7 +62,9 @@ Private-Package: \
   com.ibm.ws.cdi.interfaces.jakarta;version=latest,\
   com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
   com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
+  com.ibm.ws.container.service;version=latest,\
```
Not sure why you need this dependency, I don't see any reference to `com.ibm.ws.container.service` packages in this bundle.

abutch3r replied:
`com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl` is used to obtain the application name currently to that we know which application if multiple are installed.

And that package is part of that bundle, so it is used. if there is a better way to get the app name then we can replace it

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging30.internal_fat_tck/fat/src/io/openliberty/microprofile/reactive/messaging30/tck/ReactiveMessagingTCKLauncher.java:
```diff
@@ -48,12 +48,12 @@ public static void setUp() throws Exception {
     @Test
     @Mode(TestMode.FULL)
-    @AllowedFFDC({ "org.jboss.weld.exceptions.DeploymentException", "com.ibm.ws.container.service.state.StateChangeException" }) // The tested deployment exceptions cause FFDC so we have to allow for this.
+    @AllowedFFDC({ "org.jboss.weld.exceptions.DeploymentException", "com.ibm.ws.container.service.state.StateChangeException", "jakarta.enterprise.inject.spi.DeploymentException" }) // The tested deployment exceptions cause FFDC so we have to allow for this.
```
Not sure why you need to add `jakarta.enterprise.inject.spi.DeploymentException` - I _think_ you'd only need that if someone was creating an instance of exactly that type and throwing it, whereas now you're just rethrowing the `DeploymentException` generated by smallrye RM, and that didn't need this previously.

abutch3r replied:
Can confirm having rerun, we now do get `jakarta.enterprise.inject.spi.DeploymentException` FFDCS, coming from the smallrye code 
```
Stack Dump = jakarta.enterprise.inject.spi.DeploymentException: Wiring error(s) detected in application.
	at io.smallrye.reactive.messaging.providers.extension.MediatorManager.start(MediatorManager.java:243)
	at io.smallrye.reactive.messaging.providers.extension.MediatorManager$Proxy$_$$_WeldClientProxy.start(Unknown Source)
```

Looking at the Smallrye code, this is what we should of been getting as they do throw the `Jakarta` version, so it is fact that we are now catching this earlier in app starting process that we get the `Jakarta` FFDC as well as the WELD one.

Azquelt reviewed (commented):
A few comments but I don't think any of them should prevent this PR being merged. They can all be fixed up later.
